### PR TITLE
agent: add secure_join to prevent softlink escape

### PR DESCRIPTION
**Pull request by houstar**
This patch fixed the security issue if the container images has
    unsafe symlink to the container rootfs and hackers can be exploit
    this symlink to hack the guest system. e.g. make directory or files
    on guest.
    
    CVE-2020-2026

Signed-off-by: Qingyuan Hou <qingyuan.hou@linux.alibaba.com>

### CODE REVIEW

**Review by lifupan — APPROVED** [review]

**Comment by bergwolf** [comment]
Please open an issue to describe the problem and add `Fixes: #<issue-id>` to your commit message.

**Comment by houstar** [comment]
@bergwolf 

See CVE-2020-2026

**Comment by bergwolf** [comment]
@houstar I meant that in order to pass travis CI, you need to reformat the commit message to add a `Fixes: <issue>` line, following the contributor guide. Please see https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#closing-issues for details

**Comment by houstar** [comment]
@bergwolf  Updated, thanks

**Comment by YuvalAvra** [comment]
@houstar Hi, I'm don't think this is related to [CVE-2020-2026](https://github.com/kata-containers/community/blob/master/VMT/KCSA/KCSA-CVE-2020-2026.md), which was a malicious guest to host attack. CVE-2020-2026's [fix was applied to the kata-runtime](https://github.com/kata-containers/runtime/pull/2713). I didn't get to the bottom of this PR, but since the changes are made to the kata-agent (guest internal), I don't think it's related to CVE-2020-2026.

**Comment by bergwolf** [comment]
@houstar please explain a bit why you think it is related to CVE-2020-2026. IMO the PR looks like a fix to [CVE-2015-3629](https://nvd.nist.gov/vuln/detail/CVE-2015-3629?) instead.

CVE-2020-2026 was supposed to be fixed at the runtime side by https://github.com/kata-containers/kata-containers/commit/042135949a15cf610179c2e8a0318b34d408b177

**Comment by houstar** [comment]
See details in https://github.com/kata-containers/kata-containers/issues/1219

**Comment by houstar** [comment]
In reply to @YuvalAvra  @bergwolf 

See details in Fixes #1219 ,  make directory or file on guest is just the first step to malicious guest to host attack.

With this patch,  anyone cann't make directory or file on guest anymore.

**Comment by bergwolf** [comment]
@houstar I looked at #1219 and it looks like CVE-2015-3629, which is a container symlink escape into the host (in kata's case, the guest) file system. The PR is a good fix itself so that we can better protect the guest root image, which in Kata's case is already readonly.

CVE-2020-2026 OTOH is to trick the runtime to mount guest rootfs on any host file system and need to be fixed at the host (runtime) side.

**Comment by houstar** [comment]
Updated the CVE number, thanks

File: src/agent/rustjail/src/mount.rs [hunk 1]
```diff
@@ -1224,4 +1266,80 @@ mod tests {
...
+    #[test]
+    fn test_rootfs_absolute_softlink() {
+        let rootfs = "/tmp/rootfs4";
+        // /var/run/4 -> /
+        let unsafe_path = "/var/run/4";
+
+        // prepare softlink to test
+        remove_dir_all("/tmp/rootfs4").unwrap_or_default();
+        create_dir_all("/tmp/rootfs4/var/run").expect("failed to create rootfs");
+        fs::symlink("/", "/tmp/rootfs4/var/run/4").unwrap();
```
Comment (jodh-intel) [inline src/agent/rustjail/src/mount.rs:1341]: Please can you add appropriate calls to `skip_if_not_root()` / `skip_if_root()` for these tests.

File: src/agent/rustjail/src/mount.rs [hunk 2]
```diff
@@ -668,6 +668,44 @@ fn parse_mount(m: &Mount) -> (MsFlags, String) {
     (flags, data.join(","))
 }
 
+fn secure_join(rootfs: &str, unsafe_path: &str) -> String {
```
Comment (jodh-intel) [inline src/agent/rustjail/src/mount.rs:671]: - Please add a comment explaining what this function does.
- Please can you add a set of unit tests for this function that test all the awkward boundary conditions (`rootfs == ""`, `unsafe_path == ""`, *etc*). I'd probably use a table-driven set of tests as documented in https://github.com/kata-containers/tests/blob/master/Unit-Test-Advice.md#rust-table-driven-tests.
- There are lots of calls to `unwrap()` in this fuction which makes me wonder if we want to return `Result<String>` here.

File: src/agent/rustjail/src/mount.rs [hunk 3]
```diff
@@ -668,6 +668,44 @@ fn parse_mount(m: &Mount) -> (MsFlags, String) {
...
+                            path = PathBuf::from(rootfs.to_string());
+                        }
+                    }
+                }
+            }
+        }
+        // skip any ".."
+        if path.ends_with("..") {
+            path.pop();
+        }
```
Comment (jodh-intel) [inline src/agent/rustjail/src/mount.rs:703]: I don't think you need this check if you guarantee `path` has been canonicalized?

**Review by jodh-intel — COMMENTED** [review]
Thanks for raising @houstar. A few comments...

File: src/agent/rustjail/src/mount.rs (hunk 3 above)
Comment (houstar) [reply src/agent/rustjail/src/mount.rs:703]: There is the case that rootfs is not exists, this will be cause error if  canonicalize

File: src/agent/rustjail/src/mount.rs (hunk 1 above)
Comment (houstar) [reply src/agent/rustjail/src/mount.rs:1341]: Either root or non-root user is Ok

File: src/agent/rustjail/src/mount.rs (hunk 2 above)
Comment (houstar) [reply src/agent/rustjail/src/mount.rs:671]: OK. got it

Comment (houstar) [reply src/agent/rustjail/src/mount.rs:672]: Updated

File: src/agent/rustjail/src/mount.rs [hunk 4]
```diff
@@ -1224,4 +1267,86 @@ mod tests {
...
+    #[test]
+    fn test_rootfs_not_exists() {
+        let rootfs = "/home/rootfs";
+        assert_eq!("/home/rootfs/a/b/c", secure_join(rootfs, "a/b/c"));
+    }
+
+    #[test]
+    fn test_rootfs_relative() {
+        let rootfs = "/home/rootfs";
+        assert_eq!("/home/rootfs/a/b/c", secure_join(rootfs, "../../../a/b/c"));
```
Comment (jodh-intel) [inline src/agent/rustjail/src/mount.rs:1279]: Could you add a few more test cases for `insecure_path` values of:

- `/`
- `//`, `//////`, `/../..////../..////../` (you get the idea ;)
- `/../`
- `/../../etc/`

Might be worth considering the following `root`-only tests too?

- `/proc/1/root/`
- `/proc/1/root/../`
- `/proc/1/root/../etc/`

I suggested the table-based approach as it makes adding new unit tests much easier - you don't need to keep inventing new test function names either ;)

File: src/agent/rustjail/src/mount.rs [hunk 5]
```diff
@@ -668,6 +668,45 @@ fn parse_mount(m: &Mount) -> (MsFlags, String) {
     (flags, data.join(","))
 }
 
+// Use secure_join to get full path of container rootfs
```
Comment (jodh-intel) [inline src/agent/rustjail/src/mount.rs:671]: Paths are awkward and handling can be error prone so please can you add a little more detail here? Something like:

```rust
// This function constructs a canonicalized path by combining the `rootfs` and `unsafe_path` elements.
// The resulting path is guaranteed to be ("below" / "in a directory under") the `rootfs` directory.
//
// Parameters:
//
// - `rootfs` is the absolute path to the root of the containers root filesystem directory which must exist.
// - `unsafe_path` is path inside a container. It is unsafe since it may try to "escape" from the containers
//    rootfs by using one or more "../" path elements.
```

Also, needs explanation of what happens if `unsafe_path` does not exist (what is returned?)

File: src/agent/rustjail/src/mount.rs [hunk 6]
```diff
@@ -1224,4 +1267,86 @@ mod tests {
...
+    #[test]
+    fn test_rootfs_relative_softlink_1() {
+        let rootfs = "/tmp/rootfs1";
+        // /var/run/1 -> ../../../../
+        let unsafe_path = "/var/run/1";
+
+        // prepare softlink to test
+        remove_dir_all("/tmp/rootfs1").unwrap_or_default();
+        create_dir_all("/tmp/rootfs1/var/run").expect("failed to create rootfs");
+        fs::symlink("../../../../", "/tmp/rootfs1/var/run/1").unwrap();
```
Comment (jodh-intel) [inline src/agent/rustjail/src/mount.rs:1300]: These tests don't clean up - the symlink need to be removed on success. Again, table-based tests make this a one-liner with minimal duplicated code.

**Review by jodh-intel — COMMENTED** [review]
Thanks @houstar. A few further comments...

File: src/agent/rustjail/src/mount.rs (hunk 5 above)
Comment (houstar) [reply src/agent/rustjail/src/mount.rs:671]: Thanks for point it out in advance. I'll update it later.

File: src/agent/rustjail/src/mount.rs (hunk 6 above)
Comment (houstar) [reply src/agent/rustjail/src/mount.rs:1300]: Ok. Will be updated to table-based tests

Comment (houstar) [reply src/agent/rustjail/src/mount.rs]: Done

File: src/agent/rustjail/src/mount.rs [hunk 7]
```diff
@@ -1224,4 +1274,81 @@ mod tests {
...
+            TestData {
+                name: "abs softlink points to the root",
+                rootfs: rootfs_path,
+                unsafe_path: "3",
+                symlink_path: "/",
+                result: &format!("{}/", rootfs_path).as_str().to_owned(),
+            },
+        ];
+
+        for t in tests.iter() {
```
Comment (jodh-intel) [inline src/agent/rustjail/src/mount.rs:1345]: To make debugging easier, you might want to copy'n'paste more of the template in the [unit test advice doc](https://github.com/kata-containers/tests/blob/master/Unit-Test-Advice.md#rust-table-driven-tests) since that version adds the test number and test details on failure (using the second argument to `assert!()`).

**Review by jodh-intel — APPROVED** [review]
Thanks @houstar.

File: src/agent/rustjail/src/mount.rs (hunk 7 above)
Comment (houstar) [reply src/agent/rustjail/src/mount.rs:1345]: I don't understand why this make debugging easier, could you help me to understand, thanks ?

Comment (jodh-intel) [reply src/agent/rustjail/src/mount.rs:1345]: If one of those table-based tests failed, you'll get an error showing the details of *which* test failed. Something like this:

```
thread 'main' panicked at 'test[4]: TestData { str: "foo", num: 42, result: Ok("fail-on-purpose-99") }, ...
                                ~   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                ^                               ^
                                |                               |
              From tests.iter().enumerate()   The table entry for the failing test
```

**Comment by fidencio** [comment]
I'm really sorry for the disruption but in order to do the branch name changes announced [here](http://lists.katacontainers.io/pipermail/kata-dev/2021-January/001695.html) I'm adding the "do-not-merge" and the "needs-branch-adjustment" labels to the PR.

I'll add a new comment once the branch changes are done and once it happens I'd like to ask you to resubmit the PR against the new `main` branch, possibly linking to this PR in case you find the conversations / reviews already made important to proceed.

**Comment by fidencio** [comment]
Kata Containers repos are back to business.  Thanks a lot for your patience and I'm very happy to announce that there's no changes needed from your side as renaming the branches automagically reflected here.

If you notice something weird, some missing CI, please, do not hesitate on contacting us here or on #kata-dev on bit.ly/katacontainersslack.

Also, this is a good opportunity to have your PR rebased atop of `main`, to ensure your PR follow all the guidelines (https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format) enforced by this community, or even to re-evaluate whether the PR is currently needed.

Last but not least,  as the default branch has been renamed to `main`, if you have a local clone you'd need to update it and you do so by running:
```
git branch -m 2.0-dev main
git fetch origin
git branch -u origin/main main
```

Happy hacking!

File: src/agent/rustjail/src/mount.rs (hunk 7 above)
Comment (houstar) [reply src/agent/rustjail/src/mount.rs]: Done. sorry for that exhausted by the work

**Comment by houstar** [comment]
@lifupan  @jodh-intel  @bergwolf  Could you help to merge into main branch since this patch was carefully reviewed?

**Comment by houstar** [comment]
@jodh-intel  Could you take a look the two failed CI test? it seems that not this PR related.

**Comment by egernst** [comment]
restarting failed CI job -- looked like a networking/intermittent issue when pulling an image...
